decoder/core/ndecoder.py

```
@@ -0,0 +1,309 @@
+import numpy as np 
+import pandas as pd 
+from KDEpy import FFTKDE 
+from scipy.interpolate import interp1d 
+from sklearn.model_selection import StratifiedKFold
+from decoder.core.model import Model
+import pickle
+        
+class NDecoder:
+    def __init__(self, bandwidth, min_ISIs, conditions):
+        self.bw = bandwidth
+        self.min_ISIs = min_ISIs
+        self.conditions = conditions #list of possible labels
+        self.model = Model(conditions) 
+
+    def _create_condition_subset_mapping(self, X, y):
+        all_conditions = self.conditions
+        condition_subset_mapping = {}
+
+        y = np.asarray(y)
+
+        # Check for invalid labels in y
+        invalid_labels = set(y) - set(all_conditions)
+        if invalid_labels:
+            raise ValueError(f"Invalid labels found: {', '.join(invalid_labels)}")
+
+        for condition in all_conditions:
+            condition_indices = []
+            if len(y) > 0:
+                condition_indices = np.where(y == condition)[0]
+            condition_subset = X[condition_indices]
+            condition_subset_mapping[condition] = condition_subset
+
+        return condition_subset_mapping
+
+    @staticmethod
+    def estimate_isi_distribution(log_ISIs, bandwidth):
+        """
+        Estimate the Inter-Spike Interval (ISI) Distribution using Kernel Density Estimation (KDE).
+
+        Args:
+            log_ISIs (numpy.ndarray): An array of log-transformed Inter-Spike Interval (ISI) values.
+            bandwidth (float): The bandwidth parameter for the KDE.
+
+        Returns:
+            tuple: A tuple containing two functions:
+                - A function that interpolates the estimated probability density function (PDF).
+                - A function to generate an inverse Cumulative Distribution Function (CDF) for sampling.
+        """
+        # x = np.linspace(-2, 6, 100)
+        # TODO: write x in relation to fftkde data's min and max
+        x = np.linspace(-2, 20, 100)
```

How do FFTKDE and interp1d handle values outside the data range?

---

In FFTKDE's evaluate function, it raises a ValueError if any points passed in are outside of the data range (the error message is "Every data point must be inside of the grid."). 
interp1d also throws a ValueError, but there is a parameter `bounds_error` that you can set to False so that interp1d just sets the out-of-range values with some `fill_value` (NaN by default) instead of throwing an error.

---

Hmm, I think we want the range to have some padding that goes beyond the data range (KDE should naturally be able to handle values outside the observed range). That said, we should let it throw and error and handle that error in the decoding step.